tik/tikSwap/tikSwap.cpp

```diff
@@ -88,196 +77,317 @@ int main(int argc, char *argv[])
+                                alloc->moveBefore(insertion);
+                                mappedExports[IDToValue[key.second]] = alloc;
                             }
-                            KInst->setArgOperand((unsigned int)i, newArgs[(size_t)i]);
-                        }
-                        KernelInterface a(-1, -1);
-                        for (auto &j : kernel->Exits)
-                        {
-                            if (j->Index == 0)
+                            else
                             {
-                                a.Block = j->Block;
-                                a.Index = j->Index;
+                                alloc = mappedExports[IDToValue[key.second]];
                             }
-                        }
-                        auto sw = iBuilder.CreateSwitch(cast<Value>(KInst), baseBlockMap[a.Block], (unsigned int)kernel->Exits.size());
-                        for (auto &j : kernel->Exits)
-                        {
-                            if (j->Index != 0)
+                            vector<pair<Value *, Instruction *>> toReplace;
+                            // create a load and replace the old value for every use of this export
+                            for (auto use : IDToValue[key.second]->users())
                             {
-                                sw->addCase(ConstantInt::get(Type::getInt8Ty(base->getContext()), (uint64_t)j->Index), baseBlockMap[j->Block]);
+                                if (auto inst = dyn_cast<Instruction>(use))
+                                {
+                                    // skip the kernel calls
+                                    if (auto calli = dyn_cast<CallInst>(inst))
+                                    {
+                                        if (calli->getCalledFunction() == baseFuncInst)
+                                        {
+                                            continue;
+                                        }
+                                    }
+                                    // skip the exports that exist in different contexts
+                                    if (inst->getParent()->getParent() != KInst->getParent()->getParent())
+                                    {
+                                        continue;
+                                    }
+                                    // do not do memory operations in the entrance block, unless the user is a phi
+                                    if (inst->getParent() == IDToBlock[e->Block] && dyn_cast<PHINode>(inst) == nullptr)
+                                    {
+                                        continue;
+                                    }
+                                    toReplace.emplace_back(pair(IDToValue[key.second], inst));
+                                }
+                            }
+                            KInst->replaceUsesOfWith(IDToValue[key.second], alloc);
```

Comment:
Just a comment, I've found uses of this function to be problematic. If it is working great, but it has caused numerous bugs for me in the past. Just keep an eye on it.

Reply:
haven't had a problem yet, I use it to target specific values that don't fit well into the VMap. Same case in CartographerKernel. When replacing the values of exports, we have a pointer mapping to its data. this can't just be mapped 1:1 in the VMap because the types don't match, and what I found was that the Remap() function changes the types of instructions when you remap with a different type, leading to problems.